device/cuda/src/clusterization/clusterization_algorithm.cu

```diff
@@ -92,16 +123,51 @@ clusterization_algorithm::output_type clusterization_algorithm::operator()(
     assert(m_config.max_cells_per_thread <=
            device::details::CELLS_PER_THREAD_STACK_LIMIT);
 
+    vecmem::unique_alloc_ptr<unsigned int[]> disjoint_set;
+    vecmem::unique_alloc_ptr<unsigned int[]> cluster_sizes;
+
+    // If we are keeping the disjoint set data structure, allocate space for it.
+    if (keep_disjoint_set) {
+        disjoint_set =
+            vecmem::make_unique_alloc<unsigned int[]>(m_mr.main, num_cells);
+        cluster_sizes =
+            vecmem::make_unique_alloc<unsigned int[]>(m_mr.main, num_cells);
+    }
+
     kernels::ccl_kernel<<<num_blocks, m_config.threads_per_partition,
                           2 * m_config.max_partition_size() *
                               sizeof(device::details::index_t),
                           stream>>>(
         m_config, cells, det_descr, measurements, cell_links, m_f_backup,
-        m_gf_backup, m_adjc_backup, m_adjv_backup, m_backup_mutex.get());
+        m_gf_backup, m_adjc_backup, m_adjv_backup, m_backup_mutex.get(),
+        disjoint_set.get(), cluster_sizes.get());
+
     TRACCC_CUDA_ERROR_CHECK(cudaGetLastError());
+    TRACCC_CUDA_ERROR_CHECK(cudaStreamSynchronize(stream));
+
+    std::optional<traccc::edm::silicon_cluster_collection::buffer>
+        cluster_data = std::nullopt;
+
+    if (keep_disjoint_set) {
+        assert(m_mr.host != nullptr);
+
+        auto num_measurements = m_copy.get().get_size(measurements);
+
+        std::vector<unsigned int> cluster_sizes_host;
+        cluster_sizes_host.resize(num_measurements);
+
+        TRACCC_CUDA_ERROR_CHECK(cudaMemcpy(
+            cluster_sizes_host.data(), cluster_sizes.get(),
+            num_measurements * sizeof(unsigned int), cudaMemcpyDeviceToHost));
+
+        cluster_data.emplace(cluster_sizes_host, m_mr.main, m_mr.host,
+                             vecmem::data::buffer_type::resizable);
```

Comment:
I don't understand what's going on here. 😦 How is this meant to fill the container?

I imagined that we would want to have the cluster information on the device. I was imagining that the `ccl_kernel` would fill this SoA container directly. Or possibly we would launch one more kernel that fills this SoA.

Right now you create the SoA container, don't set it up (with `vecmem::copy::setup`), and just return it. It's very understandable that it would not work right now. 🤔

Reply:
Perfect, adding the `setup` call fixes the problem. :+1: